docs: Add documentation for serialization feature (#172)
Quantco/dataframely: docs/sites/features/serialization.md

Reviewer: One general question: Would it make sense to briefly showcase what a serialized schema looks like? 
Serialization may also be used by users directly instead of going through read/scan/write/sink methods. E.g., I used it before to diff schemas.

| {class}`~dataframely.Collection`  | {meth}`~dataframely.Collection.write_parquet`, {meth}`~dataframely.Collection.sink_parquet` <br> {meth}`~dataframely.Collection.read_parquet`, {meth}`~dataframely.Collection.scan_parquet`     | {meth}`~dataframely.Collection.write_delta` <br> {meth}`~dataframely.Collection.read_delta`, {meth}`~dataframely.Collection.scan_delta`    |
| {class}`~dataframely.FailureInfo` | {meth}`~dataframely.FailureInfo.write_parquet`, {meth}`~dataframely.FailureInfo.sink_parquet` <br> {meth}`~dataframely.FailureInfo.read_parquet`, {meth}`~dataframely.FailureInfo.scan_parquet` | {meth}`~dataframely.FailureInfo.write_delta` <br> {meth}`~dataframely.FailureInfo.read_delta`, {meth}`~dataframely.FailureInfo.scan_delta` |

## Serialization in {class}`~dataframely.Schema`

A {class}`~dataframely.Schema` controls the contents of a single dataframe. In this case, serialization
means that we store a single dataframe in the storage backend and attach a string representation

> of the schema as metadata.

Reviewer: ```suggestion
A {class}`~dataframely.Schema` controls the contents of a single data frame. In this case, serialization
means that we store a single data frame in the storage backend and attach a string representation
of the schema as metadata.
```
nit: to be consistent

# ...or lazily
new_lf: dy.LazyFrame[MySchema] = MySchema.scan_parquet("my.parquet")

# Or deltalake eagerly
new_df: dy.DataFrame[MySchema] = MySchema.read_delta("/path/to/table")

# ...or lazily
new_lf: dy.LazyFrame[MySchema] = MySchema.scan_delta("/path/to/table")
```

The role of the metadata is that when reading, `dataframely` can internally check
if the `Schema` class we use for reading matches the stored metadata in the file.
If it does, we do not need to run validation again,
but we can infer that the data in the file already matches the schema, which saves us time.

## Serialization in {class}`~dataframely.Collection`

Serialization in collections works analogously to schemas. The only difference is that
we now have to handle multiple dataframes instead of a single one.
`dataframely` will therefore have to create multiple tables in the storage backend
(e.g. multiple parquet files, or multiple delta tables).

```python
# Any collection will work
class MyCollection:
    df1: dy.LazyFrame[MySchema1]
    df2: dy.LazyFrame[MySchema2]

collection: MyCollection = MyCollection.validate(...)

# Writes and reads work the same as for Schema, except that the argument is adapted

> # to allow for multiple dataframes,

Reviewer: ```suggestion
# to allow for multiple data frames,
```

# e.g. for parquet: Pass a directory instead of a path to a single parquet
collection.write_parquet("/path/to/directory/")
collection.read_parquet("/path/to/directory/")
collection.scan_parquet("/path/to/directory/")
```

Just as for `Schema`, metadata is stored in the backend to encode the schema information.

> This includes the schemas of the member dataframes as well as collection-level constraints.

Reviewer: ```suggestion
This includes the schemas of the member data frames as well as collection-level constraints.
```
